### DISCUSSION TITLE
Configure a priority of a cmp source

### DISCUSSION THREAD

**Discussion by alex35mil** [LazyVim/LazyVim#4473] [Q&A]
What would be the best way to configure a priority of one of the cmp sources?

In particular, I want to bump the priority of [`crates` source](https://github.com/LazyVim/LazyVim/blob/a1c3ec4cd43fe61e3b614237a46ac92771191c81/lua/lazyvim/plugins/extras/lang/rust.lua#L26) because yml lsp shows rubbish, but it has precedence over the `crates` which is what I need most of the time.

I tried to add `crates` source manually via cmp plugin spec, but it results in duplicated source in cmp config and mine gets ignored because the one added by LazyVim has a higher priority.

**Comment by dpetka2001** [comment]
See this [comment](https://github.com/LazyVim/LazyVim/issues/4414#issuecomment-2345564919) for how to remove a default LazyVim source and then add your own, so that there are no duplicates. As far as I know (and I might be wrong so you better read through the `nvim-cmp` docs as well), 2 things that determine the priority of the sources are their position in the table `sources` and then also `priority` option.

So, here in this example of LazyVim
```lua
        sources = cmp.config.sources({
          { name = "nvim_lsp" },
          { name = "path" },
        }, {
          { name = "buffer" },
        }),
```
`nvim_lsp` will have higher priority than `path`. `buffer` is in `group_index = 2`, so it will only be available when there are no `nvim_lsp` and `path` suggestions. 

So, you could either pass a very big `priority` value alongside the table of your source (I don't know if there's a default value) or make sure that you insert it in top position with `table.insert(opts.sources, 1, name_of_your_source)`, because `table.insert` by default appends the value to the end of the table and not the top. You might want to put it in position 2 if you want `nvim_lsp` to be higher priority.

**Reply by alex35mil** [reply]
@dpetka2001 Thanks for the explanation! For some reason, I have `crates` source prioritized, but rubbish from LSP still shows on top:

![CleanShot 2024-09-21 at 14 10 20@2x](https://github.com/user-attachments/assets/d1238492-5f86-4b2a-9fdf-b3feda381e50)

(Versions from `crates` are down below)

Here's the sources:

```
  sources = { {
      group_index = 1,
      name = "crates",
      option = {},
      priority = 1000
    }, {
      group_index = 1,
      name = "copilot",
      option = {},
      priority = 100
    }, {
      group_index = 1,
      name = "nvim_lsp",
      option = {}
    }, {
      group_index = 1,
      name = "path",
      option = {}
    }, {
      group_index = 2,
      name = "buffer",
      option = {}
    }, {
      group_index = 1,
      name = "snippets",
      option = {}
    }, {
      group_index = 0,
      name = "lazydev",
      option = {}
    }, {
      group_index = 1,
      name = "git",
      option = {}
    } },
```

**Reply by dpetka2001** [reply]
The other thing I could think of would be the `nvim-cmp`'s comparators. Unfortunately, I've never done this myself. Take a look at [nvim-cmp's docs](https://github.com/hrsh7th/nvim-cmp/blob/ae644feb7b67bf1ce4260c231d1d4300b19c6f30/lua/cmp/config/default.lua#L65-L79) and maybe also ask a question at `nvim-cmp`'s Discussions as I'm afraid I can't be of further help regarding this matter.

**Reply by dpetka2001** [reply]
Maybe try putting `compare.exact` first? I'm not sure what those do actually.

**Reply by dpetka2001** [reply]
I think the entries that have `~` at the end are LSP snippets. I used to have a function that I stumbled upon on some Reddit thread to deprioritize LSP snippets, so they wouldn't show up on top. Take a look at my [dotfiles here](https://github.com/dpetka2001/dotfiles/blob/b983e9e1d8de1076c618bdc1f1f2a70d8de309f9/dot_config/nvim/lua/plugins/coding.lua#L77-L91) for how I did it. Try it to see if it fits your needs.

**Comment by alex35mil** [comment]
Solved by setting `group_index` to zero:

```lua
table.insert(opts.sources, 1, {
    name = "crates",
    priority = 1000,
    group_index = 0,
})
```

Thanks for help @dpetka2001!
